src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -86,5 +114,255 @@ private protected IReadOnlyList<PackageReference> GetTransitivePackageReferences
+        /// </summary>
+        /// <param name="ct">Cancellation token for async operation</param>
+        /// <returns>A list of dependencies, indexed by framework/RID</returns>
+        /// <remarks>Assets file reading occurs in a background thread</remarks>
+        /// <seealso cref="GetAssetsFilePathAsync"/>
+        protected async ValueTask<IList<LockFileTarget>> GetTargetsListAsync(CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            await TaskScheduler.Default;
+
+            string assetsFilePath = await GetAssetsFilePathAsync();
+            LockFile lockFile = LockFileUtilities.GetLockFile(assetsFilePath, NullLogger.Instance);
+
+            return lockFile?.Targets;
+        }
+
+        /// <summary>
+        /// Runs Depth First Search recursively to mark current and dependend nodes with top dependency
+        /// </summary>
+        /// <param name="top">Top, Direct dependency</param>
+        /// <param name="current">Current package/node to visit</param>
+        /// <param name="graph">Package dependency graph, from assets file</param>
+        /// <param name="visited">Dictionary to remember visited nodes</param>
+        /// <param name="fxRidEntry">Framework/Runtime-ID associated with current <paramref name="graph"/></param>
+        private void MarkTransitiveOrigin(PackageReference top, PackageReference current, LockFileTarget graph, HashSet<PackageIdentity> visited, FrameworkRuntimePair fxRidEntry, CancellationToken token)
+        {
+            token.ThrowIfCancellationRequested();
+
+            LockFileTargetLibrary node = graph
+                .Libraries
+                .Where(lib => string.Equals(lib.Name, current.PackageIdentity.Id, StringComparison.OrdinalIgnoreCase)
+                        && current.AllowedVersions.Satisfies(lib.Version)
+                        && lib.Type == LibraryType.Package.Value)
+                .FirstOrDefault();
+
+            if (node != default)
+            {
+                visited.Add(current.PackageIdentity); // visited
+
+                // Update cache
+                TransitiveEntry cachedEntry = GetCachedTransitiveOrigin(current.PackageIdentity);
+
+                if (cachedEntry == null)
+                {
+                    cachedEntry = new Dictionary<FrameworkRuntimePair, IList<PackageReference>>
+                    {
+                        [fxRidEntry] = new List<PackageReference>()
+                    };
+                }
+
+                if (!cachedEntry.ContainsKey(fxRidEntry))
+                {
+                    cachedEntry[fxRidEntry] = new List<PackageReference>();
+                }
+
+                if (!cachedEntry[fxRidEntry].Contains(top))
+                {
+                    cachedEntry[fxRidEntry].Add(top);
+                }
+
+                SetCachedTransitiveOrigin(current.PackageIdentity, cachedEntry);
+
+                foreach (PackageDependency dep in node.Dependencies)
```

Comment:
From https://github.com/NuGet/NuGet.Client/pull/4119#discussion_r657557553

`node.Dependencies` is typed as `IList<PackageDependency>` 

According to comment https://github.com/NuGet/NuGet.Client/pull/4086#issuecomment-855089185 , when using `IList<T>` in a foreach loop, the compiler creates a new backing GetEumerator(), which can cause perf issues.

cc. @jebriede

Reply:
Fixed.